src/fastpath/fastpath.c

```diff
@@ -15,6 +15,339 @@
+    endpoint_t *ep_ptr;
+    tcb_t *dest;
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    word_t badge;
+    seL4_MessageInfo_t info;
+    word_t msgInfo;
+    pde_t stored_hw_asid;
+    dom_t dom;
+
+#ifdef CONFIG_KERNEL_MCS
+    handler_cap = TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbFaultHandler)->cap;
+#else
+    cptr_t handlerCPtr;
+    handlerCPtr = NODE_STATE(ksCurThread)->tcbFaultHandler;
+    handler_cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, handlerCPtr);
+#endif
+
+    if (unlikely(!cap_capType_equals(handler_cap, cap_endpoint_cap) ||
+                 !cap_endpoint_cap_get_capCanReceive(handler_cap))) {
+        vm_fault_slowpath(type);
+    }
+
+    /* Get the endpoint address */
+    ep_ptr = EP_PTR(cap_endpoint_cap_get_capEPPtr(handler_cap));
+
+    /* Get the destination thread, which is only going to be valid
+    * if the endpoint is valid. */
+    dest = TCB_PTR(endpoint_ptr_get_epQueue_head(ep_ptr));
+
+    /* Check that there's a thread waiting to receive */
+    if (unlikely(endpoint_ptr_get_state(ep_ptr) != EPState_Recv)) {
+        vm_fault_slowpath(type);
+    }
+
+    /* ensure we are not single stepping the destination in ia32 */
+#if defined(CONFIG_HARDWARE_DEBUG_API) && defined(CONFIG_ARCH_IA32)
+    if (unlikely(dest->tcbArch.tcbContext.breakpointState.single_step_enabled)) {
+        vm_fault_slowpath(type);
+    }
+#endif
+
+    /* Get destination thread.*/
+    newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+    /* Get vspace root. */
+    cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+    /* Ensure that the destination has a valid VTable. */
+    if (unlikely(! isValidVTableRoot_fp(newVTable))) {
+        vm_fault_slowpath(type);
+    }
+
+
+#ifdef CONFIG_ARCH_AARCH32
+    /* Get HW ASID */
+    stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
+    /* borrow the stored_hw_asid for PCID */
+    stored_hw_asid.words[0] = cap_pml4_cap_get_capPML4MappedASID_fp(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_IA32
+    /* stored_hw_asid is unused on ia32 fastpath, but gets passed into a function below. */
+    stored_hw_asid.words[0] = 0;
+#endif
+
+#ifdef CONFIG_ARCH_AARCH64
+    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+    /* Get HW ASID */
+    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+    /* let gcc optimise this out for 1 domain */
+    dom = maxDom ? ksCurDomain : 0;
+    /* ensure only the idle thread or lower prio threads are present in the scheduler */
+    if (unlikely(dest->tcbPriority < NODE_STATE(ksCurThread->tcbPriority) &&
+                 !isHighestPrio(dom, dest->tcbPriority))) {
+
+        vm_fault_slowpath(type);
+    }
+
+    /* Ensure that the endpoint has has grant or grant-reply rights so that we can
+    * create the reply cap */
+    if (unlikely(!cap_endpoint_cap_get_capCanGrant(handler_cap) &&
```

Comment:
should this check be done earlier, when the handler_cap is available?

Reply:
It could be, but this is where the equivalent condition is checked in the seL4_Call fastpath. I don't think it would make a difference logically if it was earlier, but I thought it would be better to keep it consistent with the existing structure of other fastpath implementations.

Reply:
While this functionality is part of the call slowpath, it's not part of the fault slowpath, so it shouldn't be part of this fastpath.  (The check at the top of this function already establishes that endpoint cap has the correct rights).